index/scorch/segment/zap/new.go

```diff
@@ -129,6 +129,54 @@ type interimLoc struct {
 	arrayposs []uint64
 }
 
+// interimDocNums is a wrapper around a roaring Bitmap, and helps as a
+// single roaring.AddRange(0, N) method call is faster than N
+// separate, sequential Add() calls.  This helps with low cardinality
+// fields (e.g., "type" field, or "gender" field), where there are
+// commons runs of sequential docNum's.  The last range of docNum's is
+// tracked as "plus 1" to leverage zero value convention.
+type interimDocNums struct {
+	bs *roaring.Bitmap // may be nil
+
+	lastRangeBegDocNumPlus1 uint64 // inclusive
+	lastRangeEndDocNumPlus1 uint64 // inclusive
+}
+
+func (d *interimDocNums) add(docNum uint64) {
+	docNumPlus1 := docNum + 1
+
+	if d.lastRangeEndDocNumPlus1 != 0 &&
+		d.lastRangeEndDocNumPlus1+1 == docNumPlus1 {
+		d.lastRangeEndDocNumPlus1 = docNumPlus1 // extend the last range
+		return
+	}
+
+	d.incorporateLastRange()
+
+	d.lastRangeBegDocNumPlus1 = docNumPlus1 // start a new last range
+	d.lastRangeEndDocNumPlus1 = docNumPlus1
+}
+
+func (d *interimDocNums) incorporateLastRange() {
+	if d.bs == nil {
+		d.bs = roaring.New()
+	}
+
+	if d.lastRangeBegDocNumPlus1 == 0 {
+		return // there's no last range
+	}
+
+	if d.lastRangeBegDocNumPlus1 == d.lastRangeEndDocNumPlus1 {
+		d.bs.Add(uint32(d.lastRangeBegDocNumPlus1 - 1))
```

Comment:
couldn't follow this -1 logic. From first glance, it looks like "lastRangeBegDocNumPlus1" was the starting of current running range right?

Reply:
the idea is docNum's normally range from zero to N...

but, here we're using zero instead mean that it's not-a-real-docNum.

so this shifts the docNum's up by one (docNumPlus1)

but, what's sorted in the roaring bitmap is a docNum, so we need to shift the docNumPlus down by 1 to get back to a docNum, so that's "docNumPlus1 - 1"